File: src/utils/follows.ts
// Followed series, persisted as ~/.config/manga-cli/follows.json.
// Local-only (no account needed): we remember the chapter count at follow/read
// time so `--updates` can show which titles have new chapters.

import { mkdir } from "node:fs/promises";
import { FOLLOWS_FILE, CONFIG_DIR } from "./paths.ts";

export interface FollowEntry {
  id: string;
  title: string;
  coverUrl?: string;
  /** Chapter count last time we checked (the baseline for "new chapters"). */
  chapterCount: number;
  followedAt: string; // ISO 8601
}

interface FollowsFile {
  follows: FollowEntry[];
}

export async function loadFollows(): Promise<FollowEntry[]> {
  try {
    const file = Bun.file(FOLLOWS_FILE);
    if (!(await file.exists())) return [];
    const data = (await file.json()) as Partial<FollowsFile>;
    return data.follows ?? [];
  } catch {
    return [];
  }
}

async function save(follows: FollowEntry[]): Promise<void> {
  try {
    await mkdir(CONFIG_DIR, { recursive: true });
    await Bun.write(FOLLOWS_FILE, JSON.stringify({ follows }, null, 2));
  } catch {
    // best-effort
  }
}

export async function isFollowed(id: string): Promise<boolean> {
  return (await loadFollows()).some((f) => f.id === id);
}

/** Add (or refresh) a follow. Returns the resulting list. */
export async function addFollow(entry: FollowEntry): Promise<void> {
  const follows = (await loadFollows()).filter((f) => f.id !== entry.id);
  follows.unshift(entry);
  await save(follows);
}

export async function removeFollow(id: string): Promise<void> {
  await save((await loadFollows()).filter((f) => f.id !== id));
}

/** Toggle follow state for a manga; returns the new state (true = now followed). */
export async function toggleFollow(entry: Omit<FollowEntry, "followedAt">): Promise<boolean> {
  if (await isFollowed(entry.id)) {
    await removeFollow(entry.id);
    return false;
  }
  await addFollow({ ...entry, followedAt: new Date().toISOString() });
  return true;
}

/** Update the stored baseline chapter count (call after checking for updates). */
export async function markSeen(id: string, chapterCount: number): Promise<void> {
  const follows = await loadFollows();
  const f = follows.find((x) => x.id === id);
  if (!f) return;
  f.chapterCount = chapterCount;
  await save(follows);
}

File: go-rewrite/internal/util/follows.go
// Followed series — same ~/.config/manga-cli/follows.json as the TS version.
// Local-only: the stored chapter count is the baseline for "new chapters".

package util

import (
	"encoding/json"
	"os"
	"time"
)

type FollowEntry struct {
	ID           string `json:"id"`
	Title        string `json:"title"`
	Source       string `json:"source,omitempty"`
	CoverURL     string `json:"coverUrl,omitempty"`
	ChapterCount int    `json:"chapterCount"`
	FollowedAt   string `json:"followedAt"`
}

type followsFile struct {
	Follows []FollowEntry `json:"follows"`
}

func LoadFollows() []FollowEntry {
	raw, err := os.ReadFile(FollowsFile)
	if err != nil {
		return nil
	}
	var f followsFile
	if json.Unmarshal(raw, &f) != nil {
		return nil
	}
	return f.Follows
}

func saveFollows(follows []FollowEntry) {
	blob, err := json.MarshalIndent(followsFile{Follows: follows}, "", "  ")
	if err != nil {
		return
	}
	if os.MkdirAll(ConfigDir, 0o755) != nil {
		return
	}
	_ = os.WriteFile(FollowsFile, blob, 0o644)
}

func IsFollowed(id string) bool {
	for _, f := range LoadFollows() {
		if f.ID == id {
			return true
		}
	}
	return false
}

func AddFollow(e FollowEntry) {
	out := []FollowEntry{e}
	for _, f := range LoadFollows() {
		if f.ID != e.ID {
			out = append(out, f)
		}
	}
	saveFollows(out)
}

func RemoveFollow(id string) {
	var out []FollowEntry
	for _, f := range LoadFollows() {
		if f.ID != id {
			out = append(out, f)
		}
	}
	saveFollows(out)
}

// ToggleFollow flips follow state; returns the new state (true = now followed).
func ToggleFollow(e FollowEntry) bool {
	if IsFollowed(e.ID) {
		RemoveFollow(e.ID)
		return false
	}
	e.FollowedAt = time.Now().UTC().Format(time.RFC3339)
	AddFollow(e)
	return true
}

// MarkSeen updates the stored baseline chapter count.
func MarkSeen(id string, count int) {
	follows := LoadFollows()
	for i := range follows {
		if follows[i].ID == id {
			follows[i].ChapterCount = count
			saveFollows(follows)
			return
		}
	}
}
